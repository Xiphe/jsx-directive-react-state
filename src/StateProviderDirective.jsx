import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import configureStore from './configureStore';

const store = configureStore();

export default function StateProviderDirective({ Elm, props, next }) {
  return (
    <Provider store={store}>
      {next(Elm, props)}
    </Provider>
  );
}

StateProviderDirective.propTypes = {
  Elm: PropTypes.node.isRequired,
  props: PropTypes.objectOf(PropTypes.any).isRequired,
  next: PropTypes.func.isRequired,
};
