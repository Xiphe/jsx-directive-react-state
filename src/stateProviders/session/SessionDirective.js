import { connect } from 'react-redux';

const selectorsMap = new WeakMap();
const nothing = {};

function mapStateToProps(_, ownProps) {
  if (!ownProps.options) {
    return nothing;
  }

  if (!selectorsMap.has(ownProps)) {
    if (ownProps.options.setter) {
      selectorsMap.set(ownProps, {});
    } else {
      selectorsMap.set(ownProps, (state) => {
        const { key, scope } = ownProps.options;

        if (!state[scope]) {
          throw new Error(`Can not use state "${scope}" if it's not initialized.`);
        }

        return {
          [ownProps.as || key]: state[scope][key],
        };
      });
    }
  }

  return selectorsMap.get(ownProps);
}

function getPayload(setter, value) {
  switch (setter) {
    case 'toggle':
      return undefined;
    default:
      return value.target ? value.target.value : value;
  }
}

function mapDispatchToProps(dispatch, ownProps) {
  if (!ownProps.options || !ownProps.options.setter) {
    return nothing;
  }

  const { key, scope, setter } = ownProps.options;

  return {
    [ownProps.as || setter](value) {
      const action = {
        type: `JDRS:${setter.toUpperCase()}:${scope}:${key}`,
      };
      const payload = getPayload(setter, value);

      if (payload !== undefined) {
        action.payload = payload;
      }

      dispatch(action);
    },
  };
}

function mergeProps(stateProps, dispatchProps, ownProps) {
  return {
    ...ownProps,
    props: {
      ...stateProps,
      ...dispatchProps,
      ...ownProps.props,
    },
  };
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(({ Elm, props, next }) => {
  return next(Elm, props);
});

