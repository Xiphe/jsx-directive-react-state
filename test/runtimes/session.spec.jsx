/* eslint-disable no-console */

import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import SessionDirective, { bootstrap } from '../../src/stateProviders/session';
import StateProviderDirective from '../../src/StateProviderDirective';

Enzyme.configure({ adapter: new Adapter() });

function terminator(Elm, props) {
  return <Elm {...props} />;
}

function Provider({ children }) {
  return <StateProviderDirective Elm="div" props={{}} next={() => children} />;
}

function withState({ options, as }) {
  return WrappedComponent => {
    return props => {
      return (
        <SessionDirective
          Elm={WrappedComponent}
          options={options}
          as={as}
          props={props}
          next={terminator}
        />
      );
    };
  };
}

const originalConsoleError = console.error;

function noop() {}

describe('session runtime', () => {
  afterEach(() => {
    console.error = originalConsoleError;
  });

  it('renders without error', () => {
    const SimpleDiv = withState({})('div');

    bootstrap({ foo: { type: 'session' } });

    const tree = renderer
      .create(
        <Provider>
          <SimpleDiv />
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders data from initial state', () => {
    const BazDiv = withState({
      options: { key: 'bar', scope: 'foo' },
    })(({ bar }) => {
      return <div>{bar}</div>;
    });

    bootstrap({ foo: { type: 'session', initialState: { bar: 'baz' } } });

    const tree = renderer
      .create(
        <Provider>
          <BazDiv />
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders renamed data from initial state', () => {
    const BazDiv = withState({
      options: { key: 'bar', scope: 'foo' },
      as: 'children',
    })('div');

    bootstrap({ foo: { type: 'session', initialState: { bar: 'baf' } } });

    const tree = renderer
      .create(
        <Provider>
          <BazDiv />
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('retrieves data previously set by setter', () => {
    const LoremSpan = withState({
      options: { key: 'lorem', scope: 'foo' },
      as: 'children',
    })('span');
    const IpsumButton = withState({
      options: { key: 'lorem', scope: 'foo', setter: 'set' },
    })(({ set }) => {
      return <button onClick={() => set('ipsum')}>click me</button>;
    });

    bootstrap({ foo: { type: 'session', initialState: { lorem: null } } });

    const wrapper = mount(
      <Provider>
        <div>
          <LoremSpan />
          <IpsumButton />
        </div>
      </Provider>,
    );

    expect(wrapper.find('span').prop('children')).toBe(null);
    wrapper.find('button').simulate('click');
    expect(wrapper.find('span').prop('children')).toBe('ipsum');
  });

  it('retrieves data obtained by a change handler', () => {
    const DisplayName = withState({
      options: { key: 'surname', scope: 'name' },
      as: 'children',
    })('span');
    const NameInput = withState({
      options: { key: 'surname', scope: 'name', setter: 'set' },
      as: 'onChange',
    })('input');

    bootstrap({ name: { type: 'session', initialState: { surname: '' } } });

    const wrapper = mount(
      <Provider>
        <div>
          <DisplayName />
          <NameInput />
        </div>
      </Provider>,
    );

    expect(wrapper.find('span').prop('children')).toBe('');
    wrapper.find('input').simulate('change', { target: { value: 'God' } });
    expect(wrapper.find('span').prop('children')).toBe('God');
  });

  it('can handle toggled actions', () => {
    const YesNo = withState({
      options: { key: 'yes', scope: 'onoff' },
    })(({ yes }) => {
      return <span>{yes ? 'yes' : 'no'}</span>;
    });
    const Checkbox = withState({
      options: { key: 'yes', scope: 'onoff', setter: 'toggle' },
    })(({ toggle }) => {
      return <input type="checkbox" onClick={toggle} />;
    });

    bootstrap({ onoff: { type: 'session', initialState: { yes: true } } });

    const wrapper = mount(
      <Provider>
        <div>
          <YesNo />
          <Checkbox />
        </div>
      </Provider>,
    );

    expect(wrapper.find('span').prop('children')).toBe('yes');
    wrapper.find('input').simulate('click');
    expect(wrapper.find('span').prop('children')).toBe('no');
  });

  it('throws when a unknown state is used', () => {
    const BazDiv = withState({
      options: { key: 'bar', scope: 'foo' },
      as: 'children',
    })('div');

    bootstrap({ bar: { type: 'session', initialState: { bar: 'baf' } } });

    console.error = noop;
    expect(() => {
      renderer
        .create(
          <Provider>
            <BazDiv />
          </Provider>,
        )
        .toJSON();
    }).toThrowErrorMatchingSnapshot();
  });

  it('throws when a unknown setter is used', () => {
    const Checkbox = withState({
      options: { key: 'bar', scope: 'foo', setter: 'baz' },
    })(({ baz }) => {
      return <input type="checkbox" onClick={baz} />;
    });

    bootstrap({ foo: { type: 'session', initialState: { bar: 'baf' } } });

    const wrapper = mount(
      <Provider>
        <Checkbox />
      </Provider>,
    );

    expect(() => {
      wrapper.find('input').simulate('click');
    }).toThrowErrorMatchingSnapshot();
  });

  it('has no access to non-session state types', () => {
    bootstrap({ bar: { type: 'nope', initialState: { bar: 'baf' } } });

    const BazDiv = withState({
      options: { key: 'bar', scope: 'foo' },
      as: 'children',
    })('div');

    console.error = noop;
    expect(() => {
      renderer
        .create(
          <Provider>
            <BazDiv />
          </Provider>,
        )
        .toJSON();
    }).toThrowErrorMatchingSnapshot();
  });
});
