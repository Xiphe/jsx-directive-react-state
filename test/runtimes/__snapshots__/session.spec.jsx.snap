// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`session runtime has no access to non-session state types 1`] = `"Can not use state \\"foo\\" if it's not initialized."`;

exports[`session runtime renders data from initial state 1`] = `
<div>
  baz
</div>
`;

exports[`session runtime renders renamed data from initial state 1`] = `
<div>
  baf
</div>
`;

exports[`session runtime renders without error 1`] = `<div />`;

exports[`session runtime throws when a unknown setter is used 1`] = `"Unknown state setter \\"BAZ\\""`;

exports[`session runtime throws when a unknown state is used 1`] = `"Can not use state \\"foo\\" if it's not initialized."`;
